### PR TITLE
implement Restore for DeleteStmt

### PR DESCRIPTION
**Pull request by arnkore**
1. implement Restore for DeleteStmt;
2. fix [issue#188](https://github.com/pingcap/parser/issues/188) support parse tbl_name.* in delete multiple-table syntax;
pingcap/tidb#8532

### CODE REVIEW

File: ast/dml.go [hunk 1]
```diff
@@ -1408,7 +1408,91 @@ type DeleteStmt struct {
...
+	if n.Quick {
+		ctx.WriteKeyWord("QUICK ")
+	}
+	if n.IgnoreErr {
+		ctx.WriteKeyWord("IGNORE ")
+	}
+
+	if n.IsMultiTable { // Multiple-Table Syntax
+		if n.BeforeFrom {
+			//   DELETE [LOW_PRIORITY] [QUICK] [IGNORE]
```
Comment (leoppro) [inline ast/dml.go]: remove L1425-L1428

File: ast/dml.go [hunk 2]
```diff
@@ -1408,7 +1408,91 @@ type DeleteStmt struct {
...
+			if err := n.Tables.Restore(ctx); err != nil {
+				return errors.Annotate(err, "An error occurred while restore DeleteStmt.Tables")
+			}
+
+			ctx.WriteKeyWord(" FROM ")
+			if err := n.TableRefs.Restore(ctx); err != nil {
+				return errors.Annotate(err, "An error occurred while restore DeleteStmt.TableRefs")
+			}
+		} else {
+			//   DELETE [LOW_PRIORITY] [QUICK] [IGNORE]
```
Comment (leoppro) [inline ast/dml.go]: ditto

File: ast/dml.go [hunk 3]
```diff
@@ -1408,7 +1408,91 @@ type DeleteStmt struct {
...
+		}
+
+		if n.Where != nil {
+			ctx.WriteKeyWord(" WHERE ")
+			if err := n.Where.Restore(ctx); err != nil {
+				return errors.Annotate(err, "An error occurred while restore DeleteStmt.Where")
+			}
+		}
+	} else { // Single-Table Syntax
+		//   DELETE [LOW_PRIORITY] [QUICK] [IGNORE] FROM tbl_name
```
Comment (leoppro) [inline ast/dml.go]: ditto

File: ast/dml.go (hunk 1 above)
Comment (arnkore) [reply ast/dml.go]: ok

File: ast/dml.go (hunk 2 above)
Comment (arnkore) [reply ast/dml.go]: ok

File: ast/dml.go (hunk 3 above)
Comment (arnkore) [reply ast/dml.go]: ok

File: ast/dml.go [hunk 4]
```diff
@@ -1408,7 +1408,87 @@ type DeleteStmt struct {
...
+				return errors.Annotate(err, "An error occurred while restore DeleteStmt.TableRefs")
+			}
+		}
+
+		if n.Where != nil {
+			ctx.WriteKeyWord(" WHERE ")
+			if err := n.Where.Restore(ctx); err != nil {
+				return errors.Annotate(err, "An error occurred while restore DeleteStmt.Where")
+			}
+		}
```
Comment (kennytm) [inline ast/dml.go]: You could move this (and also the ORDER BY and LIMIT clauses) outside of the `if n.IsMultiTable { ... } else { ... }` to de-duplicate the code.

File: parser.y [hunk 5]
```diff
@@ -4363,6 +4363,10 @@ TableName:
 	{
 		$$ = &ast.TableName{Schema:model.NewCIStr($1),	Name:model.NewCIStr($3)}
 	}
+|   Identifier '.' '*'
```
Comment (kennytm) [inline parser.y]: Please indent with tabs.

File: ast/dml.go (hunk 4 above)
Comment (arnkore) [reply ast/dml.go]: fix

File: parser.y (hunk 5 above)
Comment (arnkore) [reply parser.y]: fix

File: parser_test.go [hunk 6]
```diff
@@ -459,10 +459,60 @@ func (s *testParserSuite) TestDMLStmt(c *C) {
...
+		{"DELETE FROM t1,t2,t3 USING t1,t2,t3 where t3.a = 1", true, "DELETE FROM `t1`,`t2`,`t3` USING ((`t1`) JOIN `t2`) JOIN `t3` WHERE `t3`.`a`=1"},
+		{"DELETE FROM t2,t3 USING t1,t2,t3 where t1.a = 1", true, "DELETE FROM `t2`,`t3` USING ((`t1`) JOIN `t2`) JOIN `t3` WHERE `t1`.`a`=1"},
+		{"DELETE FROM t2.*,t3.* USING t1,t2,t3 where t1.a = 1", true, "DELETE FROM `t2`,`t3` USING ((`t1`) JOIN `t2`) JOIN `t3` WHERE `t1`.`a`=1"},
+		{"DELETE FROM t1,t2.*,t3.* USING t1,t2,t3 where t1.a = 1", true, "DELETE FROM `t1`,`t2`,`t3` USING ((`t1`) JOIN `t2`) JOIN `t3` WHERE `t1`.`a`=1"},
+
+		// for delete statement
+		{"DELETE t1, t2 FROM t1 INNER JOIN t2 INNER JOIN t3 WHERE t1.id=t2.id AND t2.id=t3.id;", true, "DELETE `t1`,`t2` FROM (`t1` JOIN `t2`) JOIN `t3` WHERE `t1`.`id`=`t2`.`id` AND `t2`.`id`=`t3`.`id`"},
+		{"DELETE FROM t1, t2 USING t1 INNER JOIN t2 INNER JOIN t3 WHERE t1.id=t2.id AND t2.id=t3.id;", true, "DELETE FROM `t1`,`t2` USING (`t1` JOIN `t2`) JOIN `t3` WHERE `t1`.`id`=`t2`.`id` AND `t2`.`id`=`t3`.`id`"},
+		// for optimizer hint in delete statement
+		{"DELETE /*+ TiDB_INLJ(t1, t2) */ t1, t2 from t1, t2 where t1.id=t2.id;", true, ""},
```
[the remark this answers is not in the document]
Comment (arnkore) [reply parser_test.go]: fix

**Comment by leoppro** [comment]
@kennytm PTAL

**Pull request merged**